src/libraries/System.Net.Security/src/System/Net/Security/SslStreamCertificateContext.cs

```diff
@@ -0,0 +1,58 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Security.Cryptography.X509Certificates;
+
+namespace System.Net.Security
+{
+    public class SslStreamCertificateContext
+    {
+        internal readonly X509Certificate2 Certificate;
+        internal readonly X509Certificate2[]? IntermediateCertificates;
+
+        public static SslStreamCertificateContext Create(X509Certificate2 target, X509Certificate2Collection? additionalCertificates, bool offline = false)
+        {
+            if (!target.HasPrivateKey)
+            {
+                throw new NotSupportedException(SR.net_ssl_io_no_server_cert);
+            }
+
+            X509Certificate2[]? intermediates = null;
+
+            using (X509Chain chain = new X509Chain())
+            {
+                if (additionalCertificates != null)
+                {
+                    foreach (X509Certificate cert in additionalCertificates)
+                    {
+                        chain.ChainPolicy.ExtraStore.Add(cert);
+                    }
+                }
+
+                chain.ChainPolicy.VerificationFlags = X509VerificationFlags.AllFlags;
+                chain.ChainPolicy.RevocationMode = X509RevocationMode.NoCheck;
+                chain.ChainPolicy.DisableCertificateDownloads = offline;
+                chain.Build(target);
+
+                if (chain.ChainElements.Count > 2)
+                {
+                    int count = chain.ChainElements.Count;
+                    intermediates = new X509Certificate2[count];
+                    for (int i = 0; i < count; i++)
+                    {
+                        intermediates[i] = chain.ChainElements[i + 1].Certificate;
+                    }
+                }
+            }
```

Comment:
```suggestion

                // Dispose the copy of the target cert.
                chain.ChainElements[0].Certificate.Dispose();
                
                // Dispose the last cert, if we didn't include it.
                for (int i = count + 1; i < chain.ChainElements.Count; i++)
                {
                    chain.ChainElements[i].Certificate.Dispose();
                }
            }
```

Reply:
if the chain is not complete, should se sent out as much as we can? e.g. the leave as well?

Reply:
The leaf is always sent :smile:

Reply:
I guess bad terminology. I meant certificate closest to the root if root is missing. e.g. is certificate `c` is signed by `i1` -> `i2` -> `i3` -> `root` and `root` is missing, send cert + i1 + i2 + i3.

Reply:
Yeah, if the last cert isn't self-signed, then all the certs after the end-entity should be sent; that's why I checked for PartialChain (it's returned when the last cert that the builder could find wasn't self-issued).